packages/compartment-mapper/src/import-hook.js

```diff
@@ -399,3 +411,292 @@ export const makeImportHookMaker = (
+    sourceMapHook = undefined,
+    exitModuleImportNowHook,
+  },
+) {
+  // Set of specifiers for modules (scoped to compartment) whose parser is not
+  // using heuristics to determine imports.
+  /** @type {Map<string, Set<string>>} compartment name ->* module specifier */
+  const strictlyRequired = new Map();
+
+  /**
+   * @param {string} compartmentName
+   */
+  const strictlyRequiredForCompartment = compartmentName => {
+    let compartmentStrictlyRequired = strictlyRequired.get(compartmentName);
+    if (compartmentStrictlyRequired !== undefined) {
+      return compartmentStrictlyRequired;
+    }
+    compartmentStrictlyRequired = new Set();
+    strictlyRequired.set(compartmentName, compartmentStrictlyRequired);
+    return compartmentStrictlyRequired;
+  };
+
+  /**
+   * @type {ImportNowHookMaker}
+   */
+  const makeImportNowHook = ({
+    packageLocation,
+    packageName: _packageName,
+    entryCompartmentDescriptor,
+    parse,
+    compartments,
+  }) => {
+    const compartmentDescriptor = compartmentDescriptors[packageLocation] || {};
+
+    // this is necessary, because there's nothing that prevents someone from calling this function with the entry compartment.
+    assert(
+      compartmentDescriptor !== entryCompartmentDescriptor,
+      'Cannot create an ImportNowHook for the entry compartment',
+    );
+
+    // this is not strictly necessary, since the types should handle it.
+    assert(
+      parse[Symbol.toStringTag] !== 'AsyncFunction',
+      'async parsers are unsupported',
+    );
+
+    packageLocation = resolveLocation(packageLocation, baseLocation);
+    const packageSources = sources[packageLocation] || create(null);
+    sources[packageLocation] = packageSources;
+    const {
+      modules:
+        moduleDescriptors = /** @type {Record<string, ModuleDescriptor>} */ (
+          create(null)
+        ),
+    } = compartmentDescriptor;
+    compartmentDescriptor.modules = moduleDescriptors;
+
+    let { policy } = compartmentDescriptor;
+    policy = policy || create(null);
+
+    // associates modules with compartment descriptors based on policy
+    // which wouldn't otherwise be there
+    if ('packages' in policy && typeof policy.packages === 'object') {
+      for (const [packageName, packagePolicyItem] of entries(policy.packages)) {
+        if (
+          !(packageName in compartmentDescriptor.modules) &&
+          packageName in compartmentDescriptor.scopes &&
+          packagePolicyItem
+        ) {
+          compartmentDescriptor.modules[packageName] =
+            compartmentDescriptor.scopes[packageName];
+        }
+      }
+    }
+
+    const { maybeReadSync, isAbsolute } = readPowers;
+
+    /** @type {ImportNowHook} */
+    const importNowHook = moduleSpecifier => {
+      if (isAbsolute(moduleSpecifier)) {
+        let shouldCallExitModuleImportNowHook = true;
+        for (const [someDescriptorName, someDescriptor] of entries(
+          compartmentDescriptors,
```

Comment:
not all of these must make sense, I'm "thinking aloud": 

1. `compartmentDescriptors` are not changing between calls to importNowHook, so most of this work could be hoisted out to the topmost maker function and only a lookup in a reindexed structure would be needed. On the other hand, the memory needed for that would be wasted most of the time as it's a very rare situation

2. I'd consider extracting it to a separate function `findCompartmentContainingAbsolutePath`

3. since this is not causing a module to be loaded, only redirecting - the further steps are going through a policy check anyway so skipping attenuators compartment in the lookup should be unnecessary as if any attenuators exist, a policy exists as well to prevent their code from being loaded by the compartment. 

4. I'm not sure if this makes sense, but could policy be used to narrow down the search here? Take only the packages allowed by policy and look in their descriptors only.

Reply:
Please create an issue and link to it from a code comment here. It's an optimization we'll definitely want.

Reply:
yaeh, maybe that issue can be closed?  [#2466](https://github.com/endojs/endo/issues/2466)